### ISSUE TITLE
React warnings

### ISSUE THREAD

**Issue by avanderhoorn** [wesbos/Learn-Redux-Starter-Files#6]
I'm getting the following errors on from the following lines
```
// Errors
warning.js:44 Warning: Main: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://fb.me/react-special-props)

warning.js:44 Warning: Main: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://fb.me/react-special-props)

// Generated from
{React.cloneElement(this.props.children, this.props)}
```

Just wondering if you have any ideas.

**Comment by silverlight513** [comment]
+1

I also get this issue

**Comment by DiegoCardoso** [comment]
I am having the same issue. Also, the React Router seems to not be working correctly as when I click on the back button, the page reloads.

**Comment by silverlight513** [comment]
I did some quick digging around and found a stack overflow question which seems to solve the errors.

Simply replace the line 
```
{React.cloneElement(this.props.children, this.props)}
```
with
```       
 {React.Children.map(this.props.children, (child) => React.cloneElement(child,{doSomething:this.doSomething}))}
```


I haven't spent much time testing it out but it seems to work and gets rid of those errors.

**Comment by sergiokopplin** [comment]
Not working here. @silverlight513 can you help us more?

**Comment by silverlight513** [comment]
@sergiokopplin, if you're using the exact same code as Wes Bos does then that should be all you need to fix the errors in your console. 

Are you getting the same errors as the ones @avanderhoorn is getting?

**Comment by sergiokopplin** [comment]
Yes, using the same code and getting the same erros as @avanderhoorn.

```
Warning: Main: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://fb.me/react-special-props)
```

```
Warning: Main: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://fb.me/react-special-props)
```

**Comment by silverlight513** [comment]
@sergiokopplin, I'm not sure what could be wrong with your code as that change should remove the errors. Could you checkout to a clean installation, use the code from the 08 directory and add use the solution I gave above. The solution applies to line 11 of Main.js.

It might be something else that's causing the errors for you.

**Comment by sergiokopplin** [comment]
I'll try. Thank you!

**Comment by silverlight513** [comment]
@sergiokopplin, have you got it working yet or are you still getting errors?

**Comment by sergiokopplin** [comment]
Still getting. It doesn't break the application but keeps warning me about in the console.

**Comment by silverlight513** [comment]
I'm looking into it now and I have been the stupidest person ever and not actually tried the fix with passing props. 

I'm working on a proper fix now and I'll get back to you shortly

**Comment by silverlight513** [comment]
Turns out this is a bug with react. The issue can be found here - https://github.com/facebook/react/pull/6268

A PR has been merged into the master branch of react (https://github.com/facebook/react/pull/6880) but no release has been made yet.

I we'll just have to put up with the warnings for now and wait until there is a new release or revert to an older release.

(Current React release is - v15.1.0)

**Comment by avanderhoorn** [comment]
If it helps anyone else at the moment, this was my workaround:
```
return (
    <div>
        {React.cloneElement(this.props.children, { ...this.props, key: undefined, ref: undefined })}
    </div>
);
```

**Comment by wesbos** [comment]
This is a weird one - I never got the warning but I'm getting tons of emails about it.

It makes sense -you shouldn't pass down key and ref as they are unique to that element. 

I'll see if I can find a better solution, but for now use what @avanderhoorn has posted

**Comment by sjmcdowall** [comment]
Well I can certainly confirm the patch @avanderhoorn posted removed the warnings .. nice!!

**Comment by to6esko** [comment]
If it helps anyone else at the moment, this was my workaround:

return (
    <div>
        {React.cloneElement(this.props.children, { ...this.props, key: undefined, ref: undefined })}
    </div>
);

This works for me. Thanks @avanderhoorn.

**Comment by nicolas-schmitt** [comment]
@wesbos 
> This is a weird one - I never got the warning but I'm getting tons of emails about it.

This warning has been introduced in [react 15](https://facebook.github.io/react/blog/2016/04/07/react-v15.html)
> Elements will now warn when attempting to read ref and key from the props.

So maybe you worked with an earlier version ?

**Comment by danillouz** [comment]
You don't have to explicitly state that `key` and `ref` are `undefined`, this works as well:

```
return (
  <div>
    {
      React.cloneElement(this.props.children, { ...this.props })
    }
  </div>
);
```

**Comment by avanderhoorn** [comment]
A quirk with this approach of cloning that I just found is when you have a hierarchy of nested children... 
>If you do `{React.cloneElement(this.props.children, { ...this.props })}`, you are effectively passing children as a prop to the children themselves, thus creating an infinite hierarchy. Why do you need `React.cloneElement()` there? You can render `this.props.children` directly. If you really need to pass all props down (Why? I don’t see you using them), you can do something like `var { children, ...rest } = this.props;` and then `React.cloneElement(children, rest)`. I still don’t understand why though.

 – Dan Abramov @gaearon
http://stackoverflow.com/questions/35640296/render-component-of-nested-routes-reactjs

I did the following and it "works" but is really ugly... It's an alternative to Dan's approach but shows the problem
```
{React.cloneElement(this.props.children, { ...this.props, children: this.props.children.props.children })}
```

**Comment by dillonraphael** [comment]
I get this error Uncaught TypeError: Cannot read property 'props' of undefined

**Comment by sjmcdowall** [comment]
@dillonraphael -- where is the exact source code line that contains the error?  Maybe you have a typo?

**Comment by dillonraphael** [comment]
I was using @avanderhoorn line. {React.cloneElement(this.props.children, { ...this.props, children: props.children.props.children })}

**Comment by dillonraphael** [comment]
@sjmcdowall I also posted something on stackoverflow http://stackoverflow.com/questions/37664139/props-is-undefined-inside-react-component

**Comment by sjmcdowall** [comment]
I used the other version : 

{React.cloneElement(this.props.children, { ...this.props, key: undefined, ref: undefined })}

Which seemed a bit cleaner and worked for me w/o any problems ..

**Comment by dillonraphael** [comment]
@sjmcdowall different error now. seems like it's getting closer! This is what pops up in browser console `Uncaught Invariant Violation: Objects are not valid as a React child (found: object with keys {id, title, type}). If you meant to render a collection of children, use an array instead or wrap the object using createFragment(object) from the React add-ons. Check the render method of `StrainGrid`.`

**Comment by dillonraphael** [comment]
@sjmcdowall weird. this worked no problem! I swear I thought I tried this.

**Comment by sjmcdowall** [comment]
@dillonraphael   Well -- upward and onward!  Glad it worked!

**Comment by samassango** [comment]
On the Single.js in this line\
const i = this.props.posts.findIndex((post) => post.code === this.props.params.postId);
when I console.log I get -1 and the issue is here in this part  "this.props.params.postId" is returning this **${post.code}**

**Comment by samassango** [comment]
Help please

**Comment by nicolas-schmitt** [comment]
Check your Link components in Photo.js, you most likely used a quote instead of a backtick.
A quick look on [es6 template strings](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals) might be useful as well.

**Comment by samassango** [comment]
Thanks problem solved
but now I get this one from Comments.js
Comments.js:19 Uncaught TypeError: Cannot read property 'map' of undefined

**Comment by escapedcat** [comment]
@samassango 
Maybe posting your full `Comments.js` or a [link like this](https://github.com/wesbos/Learn-Redux-Starter-Files/blob/master/19/components/Comments.js#L19) will make it easier to help you. 
Uploading your current version to Github might be even better.

**Comment by samassango** [comment]
@escapedcat
Thanks I solved it

**Comment by samassango** [comment]
The issue was I was naming on my Single.js and Comments.js, so you have to be check spelling if you encounter such errors in the future.

**Comment by HMAN1911** [comment]
I have submitted a [pull request](https://github.com/wesbos/Learn-Redux-Starter-Files/pull/13) correcting the warnings about refs and key.

https://github.com/wesbos/Learn-Redux-Starter-Files/pull/13

**Comment by jcache** [comment]
FYI -- Based on this react-router [master example](https://github.com/reactjs/react-router/blob/master/examples/passing-props-to-children/app.js#L52-L54), this also works:
`{ this.props.children && React.cloneElement( this.props.children, { ...this.props } ) }`

**Comment by samassango** [comment]
I got this Error coming from React library please help

ReactDOMComponentTree.js?1a2c:105 Uncaught TypeError: Cannot read property '__reactInternalInstance$vd91j4k5xb0qsb1ydp9zfr' of null

**Comment by gauravChaturvedi** [comment]
ReactDOMComponentTree.js:105 Uncaught TypeError: Cannot read property '__reactInternalInstance$fsu1vpl3i5c3oedwnu8cwhfr' of null

Getting that error ^^ on injecting a <video> element with source.

**Comment by TroyWorks** [comment]
same as the above 2.

ReactDOMComponentTree.js:105 Uncaught TypeError: Cannot read property '__reactInternalInstance$td86329l2qrh0glyydczvvx6r' of null ReactDOMComponentTree.js:105

**Comment by gaearon** [comment]
@TroyWorks 
@gauravChaturvedi 
@samassango 

And are those errors related to this tutorial?

You are not looking at React issue tracker—this is just a third party tutorial repo.
So if there is a bug in React, you are not reporting it. 😉 

If you have a way to reproduce `Uncaught TypeError: Cannot read property '__reactInternalInstance$td86329l2qrh0glyydczvvx6r'  of null` or similar, please provide a project or a fiddle reproducing the issue with instructions to https://github.com/facebook/react/issues/6472. Just saying you also have it won’t help us fix it, unfortunately.

**Comment by TroyWorks** [comment]
yeah its from the excellent tutorial, latest node / npm 6 , latest chrome

i got past it but no clue what i did specifically, just kept goin in the tutorial
